pkg/gpu/nvidia/health_check/health_checker.go

```diff
@@ -0,0 +1,164 @@
+	glog.Info("Starting GPU Health Checker")
+
+	// Building mapping between device ID and their nvml represetation
+	count, err := nvml.GetDeviceCount()
+	if err != nil {
+		return fmt.Errorf("failed to get device count: %s", err)
+	}
+
+	glog.Infof("Found %d GPU devices", count)
+	for i := uint(0); i < count; i++ {
+		device, err := nvml.NewDeviceLite(i)
+		if err != nil {
+			return fmt.Errorf("failed to read device with index %d: %v", i, err)
+		}
+		deviceName, err := util.DeviceNameFromPath(device.Path)
+		if err != nil {
+			glog.Errorf("Invalid GPU device path found: %s. Skipping this device", device.Path)
+		}
+
+		if _, ok := hc.devices[deviceName]; !ok {
+			// we only monitor the devices passed in
+			glog.Warningf("Ignoring device %s for health check.", deviceName)
+			continue
+		}
+
+		glog.Infof("Found device %s for health monitoring. UUID: %s", deviceName, device.UUID)
+		hc.nvmlDevices[deviceName] = device
+	}
+
+	hc.eventSet = nvml.NewEventSet()
+	for _, d := range hc.nvmlDevices {
+		gpu, _, _, err := nvml.ParseMigDeviceUUID(d.UUID)
+		if err != nil {
+			gpu = d.UUID
+		}
+
+		glog.Infof("Registering device %v. UUID: %s", d.Path, d.UUID)
+		err = nvml.RegisterEventForDevice(hc.eventSet, nvml.XidCriticalError, gpu)
+		if err != nil {
+			if strings.HasSuffix(err.Error(), "Not Supported") {
+				glog.Warningf("Warning: %s is too old to support healthchecking: %v. It will always be marked healthy.", d.Path, err)
+				continue
+			} else {
+				return fmt.Errorf("failed to register device %s for NVML eventSet: %v", d.Path, err)
+			}
+		}
+	}
+
+	go func() {
+		if err := hc.listenToEvents(); err != nil {
+			glog.Errorf("GPUHealthChecker listenToEvents error: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+// chechHealth registers event with NVML to detect GPU critical errors
+func (hc *GPUHealthChecker) listenToEvents() error {
+	for {
+		select {
+		case <-hc.stop:
+			close(hc.stop)
+			return nil
+		default:
+		}
+
+		e, err := nvml.WaitForEvent(hc.eventSet, 5000)
+		if err != nil || e.Etype != nvml.XidCriticalError {
+			continue
+		}
+
+		// Ignoring application errors. GPU should still be healthy
+		// See https://docs.nvidia.com/deploy/xid-errors/index.html#topic_4
+		if e.Edata == 31 || e.Edata == 43 || e.Edata == 45 {
+			continue
+		}
+
+		if e.UUID == nil || len(*e.UUID) == 0 {
+			// All devices are unhealthy
+			glog.Errorf("XidCriticalError: Xid=%d, All devices will go unhealthy.", e.Edata)
+			for id, d := range hc.devices {
+				d.Health = pluginapi.Unhealthy
+				hc.devices[id] = d
+				hc.health <- d
+			}
+			continue
+		}
+
+		for _, d := range hc.devices {
+			// Please see https://github.com/NVIDIA/gpu-monitoring-tools/blob/148415f505c96052cb3b7fdf443b34ac853139ec/bindings/go/nvml/nvml.h#L1424
+			// for the rationale why gi and ci can be set as such when the UUID is a full GPU UUID and not a MIG device UUID.
+			uuid := hc.nvmlDevices[d.ID].UUID
+			gpu, gi, ci, err := nvml.ParseMigDeviceUUID(uuid)
```

Comment:
Same as earlier, will any of the MIG logic even work?

Reply:
this should work if `DeviceNameFromPath(migPath)` matches the device names passed in. So that's why i left this part here